### ISSUE TITLE
Problem with Binary data

### ISSUE THREAD

**Issue by chamarreddy** [net-amqp-rabbitmq/net-amqp-rabbitmq#142]
Hi,

Facing issues with binary data. The data is corrupted or receiving wrong length of bytes. I think there is some issue with Bytes alignment in case of binary data. 

The Net::Rabbitmq is working fine with bytes.

**Comment by markwellis** [comment]
Can you create a simple test to demonstrate the problem. Thanks

**Comment by chamarreddy** [comment]
Ok, Let me share a simple perl example...

**Comment by chamarreddy** [comment]
Just post some data(combination of Int and Strings) with pack and try to read the same using unpack. 

Anyway i will be sending you the example soon.

**Comment by chamarreddy** [comment]
Hi Plz find attached sample perl script to dump binary data to a queue and receive the same and unpack.

**Comment by chamarreddy** [comment]
Hi ,

Any update on this?. Did you get a chance to look into this?.

**Comment by markwellis** [comment]
Sorry, I've not had a chance to look at it yet. I don't know when someone will either. All I can say for now is someone will look at it, but I can't say when.

**Comment by eserte** [comment]
@chamarreddy Just an idea: does it work if you set `content_encoding` to something unlike `utf-8`?

**Comment by manchicken** [comment]
On #146, @eserte suggested...

> So I wonder if Net::AMQP::RabbitMQ shouldn't do assume anything about encoding or charset. Or maybe it shouldn't do anything if the content_type does not match "text/*".

I think this is a good suggestion for most things, however I do worry about how `text/*` doesn't always mean text. A quick example that comes to mind is `application/javascript`. I'm certain there are others.

**Comment by pmorch** [comment]
Ok, so here is what is going on: `RabbitMQ.xs`'s `read_message()` is trying to do UTF-8 magic that messes up binary data. If the `content_encoding` header is missing or equal to `UTF-8` (case insensitively), `SvUTF8_on(body_sv)` is called, that alters your "binary" data (see below).

# TL;DR

Just add a `{ content_encoding => 'binary' }` as a prop to `$mq->publish($channel, $routing_key, $body, $options, $props)` and you're good to go - your binary data will start working.

# Details

The problem is that `JSON`'s `encode_json` (and `YAML::XS`'s `Dump`) generate UTF-8 encoded octets, not UTF-8 characters, and so are not decoded correctly by `decode_json` because of the `SvUTF8_on(body_sv)`.

Right now, the only thing that works correctly without `content_encoding` is UTF-8 strings and binary data provided none of the octet values are > 127.

RabbitMQ.xs states that UTF-8 strings is the default. I'd like to venture that this is the wrong default.

[publish's doc](https://metacpan.org/pod/Net::AMQP::RabbitMQ#publish($channel,-$routing_key,-$body,-$options,-$props)) says:

> `$body` is the payload to enqueue.

The word "payload" means a binary blob when I read it.

Also, won't most people be sending _octets_ representing some kind of structured data through RabbitMQ? `JSON`'s `encode_json`/`decode_json` and `YAML::XS`'s `Dump`/`Load` both fail because of this. GRPC is bound to fail too. Does _anybody_ really send UTF-8 strings in a real-world application? Using some kind of home-brewed encoding? Yikes.

The only exception I can think of is `JSON`'s `to_json`/`from_json` since they use UTF-8 characters, not octets like `encode_json`/`decode_json` do. `JSON`'s documentation uses `encode_json`/`decode_json` throughout, though.

Anyway, perhaps this train has left the station. The default so far has been UTF-8 payload. Changing it would break backwards compatibility.

# Workaround

When calling publish, put in a prop:

    content_encoding => $encoding

where

    $encoding !~ /^UTF-8$/i

Then `SvUTF8_on(body_sv)` won't be called, your binary data will be left alone, and your code will start to work (as mine did... :-D).

# RabbitMQ.xs Details

RabbitMQ.xs has:

    int is_utf8_body = 1; /* The body is UTF-8 by default */
    ...
    /*
     * Since we could have UTF-8 in our content-encoding, and most people seem like they
     * treat this like the default, we're looking for the presence of content-encoding but
     * the absence of a case-insensitive "UTF-8".
     */
    if (
      strnlen(p->content_encoding.bytes, p->content_encoding.len) > 0
      &&
      (strncasecmp(p->content_encoding.bytes, "UTF-8", p->content_encoding.len) != 0)
    ) {
      is_utf8_body = 0;
    }
    ...
    if (is_utf8_body)
      SvUTF8_on(body_sv);

# I'm working on a pull-request

It'll leave the main code exactly as it is. It will update some tests in `t/*.t` (mostly `t/019_utf8_safety.t` and `t/021_json_utf8_in_and_out.t`) showing how to reliably use `JSON`'s `encode_json`/`decode_json` and showing how to get the `pack`-ed data from @chamarreddy 's sample_pack to work.

It'll update the documentation to describe how and when to use `content_encoding => 'binary'`.

Stay tuned here for more...

**Comment by manchicken** [comment]
No way to replicate this issue that I can find.
